### ISSUE TITLE
Warnings should be given in TC when names are not used

### ISSUE THREAD

**Issue by joey-coleman** [symphonytool/symphony#102]
> The following bug was [originally reported](https://sourceforge.net/p/compassresearch/bugs/102/) on Sourceforge by [pglvdm](https://sourceforge.net/u/pglvdm/profile/), 2013-10-08 11:36:39.998000:

In a model like the one attached the "result" name introduced in implicit style functions are not used so the writer thinks that the model is fine where in reality it is missing a "result =".

```
types
  Aircraft = token

channels
  init, success, failure
  givePermission, recordLanding, recordTakeOff : Aircraft
  getLanded, getPermission : set of Aircraft
  numberWaiting : int

process Airport =
begin
  state
    permission : set of Aircraft
    landed     : set of Aircraft
    inv landed subset permission
  functions
    permitted(a : Aircraft) result : bool
      post a in set permission
    down(a : Aircraft) result : bool
      post a in set landed
  operations
    Init : () ==> ()
    Init() ==
      permission := {} ; landed := {}
    GivePermission : Aircraft ==> ()
    GivePermission(a) ==
      permission := permission union {a}
      pre not permitted(a)
    RecordLanding : Aircraft ==> ()
    RecordLanding(a) ==
      landed := landed union {a}
      pre permitted(a) and not down(a)
    RecordTakeOff : Aircraft ==> ()
    RecordTakeOff(a) ==
      permission := permission \ {a} ; landed := landed \ {a}
      pre down(a)
    GetLanded: () ==> set of Aircraft
    GetLanded() ==
      return landed
    GetPermission: () ==> set of Aircraft
    GetPermission() ==
      return permission
    NumberWaiting : () ==> int
    NumberWaiting() ==
      return card (permission \ landed)
  actions
    Cycle =
      ( givePermission?a -> 
          ( [not permitted(a)] & success -> GivePermission(a)
            []
            [permitted(a)] & failure -> Skip )
        []
        recordLanding?a ->
          ( [permitted(a) and not down(a)] & success -> RecordLanding(a)
            []
            [permitted(a) => down(a)] & failure -> Skip )
        []
        recordTakeOff?a ->
          ( [down(a)] & success -> RecordTakeOff(a)
            []
            [not down(a)] & failure -> Skip )
        []
        getLanded!(GetLanded()) -> Skip
        []
        getPermission!(GetPermission()) -> Skip
        []
        numberWaiting!(NumberWaiting()) -> Skip ) ; Cycle
  @
    init -> Cycle
end
```

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/102/#b038) by [joeycoleman](https://sourceforge.net/u/joeycoleman/profile/), 2013-10-09 07:59:06.990000:
- Description has changed:

Diff:

```

--- old
+++ new
@@ -1,72 +1,72 @@
 In a model like the one attached the "result" name introduced in implicit style functions are not used so the writer thinks that the model is fine where in reality it is missing a "result =".

-types
-  Aircraft = token
+    types
+      Aircraft = token

-channels
-  init, success, failure
-  givePermission, recordLanding, recordTakeOff : Aircraft
-  getLanded, getPermission : set of Aircraft
-  numberWaiting : int
+    channels
+      init, success, failure
+      givePermission, recordLanding, recordTakeOff : Aircraft
+      getLanded, getPermission : set of Aircraft
+      numberWaiting : int

-process Airport =
-begin
-  state
-    permission : set of Aircraft
-    landed     : set of Aircraft
-    inv landed subset permission
-  functions
-    permitted(a : Aircraft) result : bool
-      post a in set permission
-    down(a : Aircraft) result : bool
-      post a in set landed
-  operations
-    Init : () ==> ()
-    Init() ==
-      permission := {} ; landed := {}
-    GivePermission : Aircraft ==> ()
-    GivePermission(a) ==
-      permission := permission union {a}
-      pre not permitted(a)
-    RecordLanding : Aircraft ==> ()
-    RecordLanding(a) ==
-      landed := landed union {a}
-      pre permitted(a) and not down(a)
-    RecordTakeOff : Aircraft ==> ()
-    RecordTakeOff(a) ==
-      permission := permission \ {a} ; landed := landed \ {a}
-      pre down(a)
-    GetLanded: () ==> set of Aircraft
-    GetLanded() ==
-      return landed
-    GetPermission: () ==> set of Aircraft
-    GetPermission() ==
-      return permission
-    NumberWaiting : () ==> int
-    NumberWaiting() ==
-      return card (permission \ landed)
-  actions
-    Cycle =
-      ( givePermission?a -> 
-          ( [not permitted(a)] & success -> GivePermission(a)
+    process Airport =
+    begin
+      state
+        permission : set of Aircraft
+        landed     : set of Aircraft
+        inv landed subset permission
+      functions
+        permitted(a : Aircraft) result : bool
+          post a in set permission
+        down(a : Aircraft) result : bool
+          post a in set landed
+      operations
+        Init : () ==> ()
+        Init() ==
+          permission := {} ; landed := {}
+        GivePermission : Aircraft ==> ()
+        GivePermission(a) ==
+          permission := permission union {a}
+          pre not permitted(a)
+        RecordLanding : Aircraft ==> ()
+        RecordLanding(a) ==
+          landed := landed union {a}
+          pre permitted(a) and not down(a)
+        RecordTakeOff : Aircraft ==> ()
+        RecordTakeOff(a) ==
+          permission := permission \ {a} ; landed := landed \ {a}
+          pre down(a)
+        GetLanded: () ==> set of Aircraft
+        GetLanded() ==
+          return landed
+        GetPermission: () ==> set of Aircraft
+        GetPermission() ==
+          return permission
+        NumberWaiting : () ==> int
+        NumberWaiting() ==
+          return card (permission \ landed)
+      actions
+        Cycle =
+          ( givePermission?a -> 
+              ( [not permitted(a)] & success -> GivePermission(a)
+                []
+                [permitted(a)] & failure -> Skip )
             []
-            [permitted(a)] & failure -> Skip )
-        []
-        recordLanding?a ->
-          ( [permitted(a) and not down(a)] & success -> RecordLanding(a)
+            recordLanding?a ->
+              ( [permitted(a) and not down(a)] & success -> RecordLanding(a)
+                []
+                [permitted(a) => down(a)] & failure -> Skip )
             []
-            [permitted(a) => down(a)] & failure -> Skip )
-        []
-        recordTakeOff?a ->
-          ( [down(a)] & success -> RecordTakeOff(a)
+            recordTakeOff?a ->
+              ( [down(a)] & success -> RecordTakeOff(a)
+                []
+                [not down(a)] & failure -> Skip )
             []
-            [not down(a)] & failure -> Skip )
-        []
-        getLanded!(GetLanded()) -> Skip
-        []
-        getPermission!(GetPermission()) -> Skip
-        []
-        numberWaiting!(NumberWaiting()) -> Skip ) ; Cycle
-  @
-    init -> Cycle
-end
+            getLanded!(GetLanded()) -> Skip
+            []
+            getPermission!(GetPermission()) -> Skip
+            []
+            numberWaiting!(NumberWaiting()) -> Skip ) ; Cycle
+      @
+        init -> Cycle
+    end

```
- **assigned_to**: Claus B. Nielsen

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/102/#0b99) by [joeycoleman](https://sourceforge.net/u/joeycoleman/profile/), 2013-10-09 07:59:07.735000:

I think this is probably a TC bug, but I think Peter needs to clarify the problem a bit (I'm not sure exactly what is wrong).

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/102/#0bf1) by [clausbn](https://sourceforge.net/u/clausbn/profile/), 2013-10-11 12:54:42.646000:

This is found by the typechecker, but the warnings are never read correctly when TC completes. 

This relates to the general issues with error reporting, and has to be fixed when the error reporting scheme is changed in the near future

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/102/#2e90) by [clausbn](https://sourceforge.net/u/clausbn/profile/), 2013-10-17 09:26:59.642000:
- **status**: open --> accepted

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/102/#b8eb) by [clausbn](https://sourceforge.net/u/clausbn/profile/), 2013-10-17 09:27:00.314000:

Warnings are now being reported in the tool. 

But is seems incorrect warnings are being given on "unused variables".

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/102/#39ed) by [lausdahl](https://sourceforge.net/u/lausdahl/profile/), 2013-11-04 14:10:20.224000:

Fixed and tested in [6a1b39a27428241974428cf29ee44819e6505ca7]

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/102/#b396) by [lausdahl](https://sourceforge.net/u/lausdahl/profile/), 2013-11-04 14:10:35.676000:
- **status**: accepted --> closed
- **Module**:  --> type-checker
